File: _589_n-ary-tree-preorder-traversal.java
// https://leetcode.com/problems/n-ary-tree-preorder-traversal/description/
public class _589_n-ary-tree-preorder-traversal {
   public List<Integer> preorder(Node root) {
        List<Integer> res = new ArrayList();
	dfs(res, root);
	return res;
    }
	
    private void dfs(List<Integer> res, Node root) {
	if (root == null) return;

	res.add(root.val);
	for (Node child : root.children) {
		dfs(res, child);
	}
    }
}

File: _589_n-ary-tree-preorder-traversal.cs
public class Solution {
    public IList<int> Preorder(Node root) {
        var results = new List<int>();
        
        if(root != null){
            results.Add(root.val);
            foreach(var child in root.children){
                results.AddRange(Preorder(child));
            }
        }
        return results;
    }
}
